### PR TITLE
Arrow dataset builder to be able to load and stream Arrow datasets

### PR DESCRIPTION
**Pull request by mariusz-jachimowicz-83**
This adds a Arrow dataset builder to be able to load and stream from already preprocessed Arrow files.
It's related to https://github.com/huggingface/datasets/issues/3035

### CODE REVIEW

File: src/datasets/packaged_modules/arrow/arrow.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+        schema = self.info.features.arrow_schema if self.info.features is not None else None
+        if self.info.features is not None and self.config.columns is not None:
+            if sorted(field.name for field in schema) != sorted(self.config.columns):
+                raise ValueError(
+                    f"Tried to load parquet data with columns '{self.config.columns}' with mismatching features '{self.info.features}'"
+                )
+        for file_idx, file in enumerate(itertools.chain.from_iterable(files)):
+            for batch_idx, record_batch in enumerate(_memory_mapped_record_batch_reader_from_file(file)):
+                pa_table = pa.Table.from_batches([record_batch])
+                yield f"{file_idx}_{batch_idx}", self._cast_table(pa_table)
```
Comment (lhoestq) [inline src/datasets/packaged_modules/arrow/arrow.py:73]:
```suggestion
            with open(file, "rb") as f:
                for batch_idx, record_batch in enumerate(pa.ipc.open_stream(f)):
                    pa_table = pa.Table.from_batches([record_batch])
                    yield f"{file_idx}_{batch_idx}", self._cast_table(pa_table)
```

File: src/datasets/packaged_modules/arrow/arrow.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+        for split_name, files in data_files.items():
+            if isinstance(files, str):
+                files = [files]
+            # Use `dl_manager.iter_files` to skip hidden files in an extracted archive
+            files = [dl_manager.iter_files(file) for file in files]
+            # Infer features is they are stoed in the arrow schema
+            if self.info.features is None:
+                for file in itertools.chain.from_iterable(files):
+                    self.info.features = datasets.Features.from_arrow_schema(read_schema_from_file(file))
+                    # with open(file, "rb") as f:
```
Comment (lhoestq) [inline src/datasets/packaged_modules/arrow/arrow.py:51]: Same here, we can't use read_schema_from_file. But using `open()` and pyarrow:
```suggestion
                    with open(file, "rb") as f:
                        self.info.features = datasets.Features.from_arrow_schema(pa.ipc.open_stream(f).schema)
```

**Review by lhoestq — COMMENTED** [review]
Nice good job !

Note: when a dataset is loaded in streaming mode, the `files` used in `_generate_tables` may point to remote files (e.g. URLs). Therefore you can't really use `_memory_mapped_record_batch_reader_from_file` which only works on local files.

Instead you can use the `open` built-in, that we actually extend to support remote files, see my suggestion. I haven't tested though, let me know if it works !

File: src/datasets/packaged_modules/arrow/arrow.py (hunk 1 above)
Comment (mariusz-jachimowicz-83) [reply src/datasets/packaged_modules/arrow/arrow.py]: Indeed. Tips applied.

File: src/datasets/packaged_modules/arrow/arrow.py (hunk 2 above)
Comment (mariusz-jachimowicz-83) [reply src/datasets/packaged_modules/arrow/arrow.py]: Indeed. Tips applied.

File: src/datasets/packaged_modules/arrow/arrow.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+            pa_table = table_cast(pa_table, self.info.features.arrow_schema)
+        return pa_table
+
+    def _generate_tables(self, files):
+        schema = self.info.features.arrow_schema if self.info.features is not None else None
+        if self.info.features is not None and self.config.columns is not None:
+            if sorted(field.name for field in schema) != sorted(self.config.columns):
+                raise ValueError(
+                    f"Tried to load parquet data with columns '{self.config.columns}' with mismatching features '{self.info.features}'"
+                )
```
Comment (lhoestq) [inline src/datasets/packaged_modules/arrow/arrow.py:69]:
```suggestion
```

**Review by lhoestq — APPROVED** [review]
Looks like it's working like a charm ! Congrats :)

Let me just remove the "columns" config parameter since it's not used anywhere and we can merge.

EDIT: and actually `batch_size` as well, since it requires calling `read_all()` which is not super efficient in streaming mode and can take some time

**Comment by lhoestq** [comment]
Let's add some documentation in a subsequent PR :)

In particular I think it's important to note to users that local arrow data are copied to cache according to the way load_dataset works, but if they want they can use Dataset.from_file instead

**Pull request merged**